docker/flux-entrypoint.sh

```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+gpg_import_dir="$HOME/gpg-import"
+
+if [ -d "$gpg_import_dir" ]; then
+    IFS="$(printf '\n')"
+    for key in "$gpg_import_dir"/*; do
+        if [ -e "$key" ]; then
+            echo "Importing GPG key: $(basename "$key")"
```

Comment:
What about calling `fluxd --version` before the key import loop? It wouldn't be output in the normal log format, though.

We could add a `--log-version` flag that does the same as `--version`, but in the log format. Hacky? Maybe.

If we actually incorporated GPG key importing into fluxd itself we could do away with this entrypoint script altogether. What are your thoughts on that? Originally I felt that this was outside of the scope of fluxd's responsibilities, but I'd be glad to discuss that further.

Reply:
> If we actually incorporated GPG key importing into fluxd itself we could do away with this entrypoint script altogether. What are your thoughts on that? Originally I felt that this was outside of the scope of fluxd's responsibilities, but I'd be glad to discuss that further.

I'm inclined to agree with you -- if you were running fluxd in another context, you'd arrange for the gpg keyring some other way.

Another alternative (I think): expect the whole keyring to be mounted, already assembled. This means more work to set it up ahead of time, but it'd be easier to update it while fluxd was running (since it would just be replaced, rather than assembled at startup). Not sure if that's a good trade-off.

If that's no good, a middle ground would be something like `--git-import-gpg-keys=/path/to/dir` (and we incorporate the script into flux code, as you mention).

Reply:
NB none of this is a condition for merging this PR; it would be good to figure it out before this feature makes its way into a release, though, since that will make the UI to it harder to change.

Reply:
> Another alternative (I think): expect the whole keyring to be mounted, already assembled. This means more work to set it up ahead of time, but it'd be easier to update it while fluxd was running (since it would just be replaced, rather than assembled at startup). Not sure if that's a good trade-off.

I considered this, but I wasn't sure if there was a **good** way to take an entire `$GNUPGHOME` directory and upload it as a Kubernetes secret. It also seems more prone to breakage if an updated version of GPG changes the format of files on disk—but I imagine that they would have to account for people having old keyrings and provide some sort of automatic-upgrade path as well.

If there's a good story for getting a whole `$GNUPGHOME` directory in a secret, and then volume-mounting it into the container, that has a lot of advantages. OTOH, I haven't been working with Kubernetes very long, but it seems like secrets are usually just tiny pieces of secret data, not whole directory trees.

We could also consider a sidecar container responsible for importing keys that writes to a shared volume used by the flux container. The keys don't need to be in place before fluxd starts up, just before fluxd tries to perform a signed commit; and, if they aren't available yet for some reason, the commit should fail and succeed on the next attempt. I'd have to flesh this out to see if it'd work like I'm thinking, but what are your initial thoughts?

Reply:
> We could also consider a sidecar container responsible for importing keys that writes to a shared volume used by the flux container. The keys don't need to be in place before fluxd starts up, just before fluxd tries to perform a signed commit; and, if they aren't available yet for some reason, the commit should fail and succeed on the next attempt. I'd have to flesh this out to see if it'd work like I'm thinking, but what are your initial thoughts?

You could put them in an `initContainer`, then it has to have run before fluxd starts up. One trickiness is choosing the container image such that there's not another bulky image to download -- one possibility is to bake the script (and gnupg) into the flux container, since we know they will have that image anyway.

Reply:
The `initContainer` solution seems like it'd work perfectly. I will try to get that done as soon as I've got some cycles!

Reply:
Aces -- I think we've puzzled it out! Thanks for your patience dealing with my comments here @scjudd, I appreciate it.